Add ArrayIndentationFixer

julienfalque opened the pull request:
Closes #770 and #1438.

SpacePossum commented on src/Fixer/Whitespace/ArrayIndentationFixer.php:
```diff
@@ -0,0 +1,331 @@
+                'start' => $arrayTokenRanges[0][0],
+                'end' => $arrayTokenRanges[count($arrayTokenRanges) - 1][1],
+                'token_ranges' => $arrayTokenRanges,
+            ];
```
maybe
```php

// remove $arrays

foreach(...) {
    $array = [
        'start' => $arrayTokenRanges[0][0],
        'end' => $arrayTokenRanges[count($arrayTokenRanges) - 1][1],
        'token_ranges' => $arrayTokenRanges,
    ];
    $array['line_braces_deltas'] = $this->computeArrayLineBracesDeltas($tokens, $array);

    yield $array;
}

// remove next foreach and return statement
```

julienfalque replied:
I removed the second `foreach` but didn't use `yield`.

EikeDawid commented:
@julienfalque are you still working on this PR or has it been abandoned? - I would love to have a consistent way to format arrays.

julienfalque commented:
The PR is not abandoned but it needs more review. If you have time, any help is welcome :)

keradus commented:
@EikeDawid, why won't you help and make a decent review? it would help merging the PR for sure

EikeDawid commented:
@keradus let me see if i can find somebody

mgarciadelojo commented:
Hi!

First of all, thank you @julienfalque for the Pull Request. Looks amazing!

I've been playing with it for a while, and everything seems fine to me. You can see a few examples below:

**Input:**

```
<?php

$array = array(
		'a' => 1,
		'asdf'      => 2,
		'adddd' => 3,
		'c' => 4,
	);


$array = array(
		 1,
		2,
		  3,
		4,
		    5,
		6
		);


$array = array(1,2,3,4,5,    );


$this->assertEquals(
    [
    'firstKey' 		=> '50',
    'secondKey' 		=> '90',
    'whateverKey' 		=> '/tmp/report',
    ],
    $this->configuration->getLoggingConfiguration()
);

return array(
  'first-name'	 => 'First Name',
  'last-name' 	 => 'Last Name',
  'language' => array(
    	'en' => 'English',
    	'es'  =>    'Spanish',
  ),
);
```

**Output:**

```
<?php

$array = [
    'a' => 1,
    'asdf' => 2,
    'adddd' => 3,
    'c' => 4,
];

$array = [
    1,
    2,
    3,
    4,
    5,
    6,
];

$array = [1, 2, 3, 4, 5];

$this->assertSame(
    [
        'firstKey' => '50',
        'secondKey' => '90',
        'whateverKey' => '/tmp/report',
    ],
    $this->configuration->getLoggingConfiguration()
);

return [
    'first-name' => 'First Name',
    'last-name' => 'Last Name',
    'language' => [
        'en' => 'English',
        'es' => 'Spanish',
    ],
];
```

The only thing I am missing are more tests, but apart from that, works great.

Cheers

keradus commented on README.rst:
```diff
@@ -247,6 +247,10 @@ Choose from the list of available rules:
 
+* **array_indentation**
+
+  Multiline arrays must be indented with configured indentation.
```
actually, I would drop ` with configured indentation` part, it's misleading, we don't have config for this fixer.

you meant global config, but that is used across many rules, and we don't mention that "global config" for them.

julienfalque replied:
Updated.

keradus commented on src/Fixer/Whitespace/ArrayIndentationFixer.php:
```diff
@@ -0,0 +1,331 @@
+
+                $previousLineBracesDelta = $lineBracesDelta;
+
+                $tokens[$index] = new Token([T_WHITESPACE, $content]);
```
just double-checking, does this fixer any time inserting new token to the collection?

julienfalque replied:
It doesn't.

keradus commented on tests/Fixer/Whitespace/ArrayIndentationFixerTest.php:
```diff
@@ -0,0 +1,563 @@
+ ];
+INPUT
+                ,
+            ],
```
it shall pass, but please add
```php
            [
                <<<'EXPECTED'
<?php
$foo = [
    'foo', // comment
    'bar',
];
EXPECTED
                ,
                <<<'INPUT'
<?php
$foo = [
  'foo', // comment
'bar',
 ];
INPUT
                ,
            ],
```

just to be sure

julienfalque replied:
Added.

julienfalque commented:
WIP as I found test cases that currently fail.

julienfalque commented:
I refactored the fixer a bit, added and fixed test cases. I feel more confident about the stability of the fixer now. I found an edge case that is still not supported though:

```diff
 $foo = [
     0 =>
-        'bar',
+    'bar',
     1 =>
-        'baz',
+    'baz',
 ];
```
IMO this case should be left untouched, i.e. values that are on their own line (rather than after `=>`) should be indented once. But I assume this is a very rare edge case (found only once in Symfony codebase), I'd say this should be a separate PR so this one can get merged. WDYT?

keradus commented:
I would say that for this case we do have `no_multiline_whitespace_around_double_arrow` rule ;)

keradus commented:
Thank you @julienfalque.

keradus commented:
really nice here !

julienfalque commented:
Really happy to see this merged, thanks! :)

mgarciadelojo commented:
Amazing, thanks guys
